### ISSUE TITLE
[CRASH] dead child 55

### ISSUE THREAD

**Issue by bcnewlin** [OpenSIPS/opensips#2216]
**OpenSIPS version you are running**

```
version: opensips 2.4.8 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 8f98bc0ab
main.c compiled on 21:11:17 Aug 10 2020 with gcc 7
```

**Crash Core Dump**

https://pastebin.com/cAJrdrGv

**Describe the traffic that generated the bug**

Not sure

**To Reproduce**

Not reproduced

**Relevant System Logs**

```
Aug 18 05:22:08 [7294] CRITICAL:core:handle_worker: dead child 55 (EOF received), pid 7265
```

**OS/environment information**
 - Operating System:  Amazon Linux 2
 - OpenSIPS installation:  git
 - other relevant information:

**Additional context**

**Comment by liviuchircu** [comment]
hey, @bcnewlin -- is there any change you could re-upload the backtrace on this one?  I remember it was something cryptic, along the lines of:

0. TLS library (crash)
1. cURL library transfer ...
2. OpenSIPS cURL library invocation

... so I couldn't make much progress.  Any other details would be very helpful, including (but not limited to):

* a way of reproducing this crash
* minimal opensips.cfg file that exhibits it
* curl library version on your Amazon Linux 2
* tls library version on your Amazon Linux 2

**Comment by bcnewlin** [comment]
@liviuchircu Unfortunately, I don't still have that backtrace. I did not realize it had timed out. I have #2258 open for ongoing crashes we are seeing, perhaps they are related? I am not good at determining from the backtrace whether 2 crashes are similar. Maybe we should just close this and investigate on that ticket?

**Comment by bcnewlin** [comment]
Here is the version info if it helps:
```
openssl-devel.x86_64               1:1.0.2k-19.amzn2.0.3            @amzn2-core
libcurl-devel.x86_64               7.61.1-12.amzn2.0.2              @amzn2-core
```

We do not have a way to reproduce, the crash occurred during heavy load. I have no way to know which call or sequence of events caused it. Also, as it occurred in our load testing environment which is heavily dependent on our custom configuration, so removing parts of the script would break the load test and we would not be able to reproduce.

I will say that we only use async HTTPS rest calls in 2 places in our script.

**Comment by liviuchircu** [comment]
Got these two links from #2258, this is the exact same crash as above (I'm 99% sure):

https://pastebin.com/LEC3b8qQ
https://pastebin.com/Y2PsGMT4

Thank you for the info - let me see if I can find a way to make progress.

**Comment by bogdan-iancu** [comment]
I agree that #2258 reports the same problem, so we could merge them and keep only one.

**Comment by bcnewlin** [comment]
@bogdan-iancu While some of the crashes reported in #2258 are definitely the same as this one, as @liviuchircu confirmed, it does seem that at least a few of the crashes in #2258 are *not* the same. I think one was referenced on another user's ticket as well. Something to do with TCP.

**Comment by bogdan-iancu** [comment]
Talking to @liviuchircu we agreed that all the crashes you reported may have been triggered by the some bug, most probability a memory a memory corruption, which makes opensips to crash in various points.
Are there any chances to run opensips with QM_MALLOC and DBG_MALLOC to enable some debugging code in the mem allocator ?

**Comment by bcnewlin** [comment]
@bogdan-iancu Yes, I'll get that enabled and report back when it recurs.

**Comment by bcnewlin** [comment]
We've had a number more crashes today with the debugging enabled. Here are the backtraces from 2 processes that crashed at the same time.

https://pastebin.com/fD2qHbC6
https://pastebin.com/55VvyTTE

**Comment by bcnewlin** [comment]
Any updates on this?

**Comment by bogdan-iancu** [comment]
my bad @bcnewlin :( . The pastes expired, could you please repost or email them to me ? Thanks !

**Comment by bcnewlin** [comment]
@bogdan-iancu I will upload new ones. I have been setting them to expire after 1 month, I'm not sure why they are being deleted so fast.

**Comment by bcnewlin** [comment]
Here is a new crash backtrace: https://pastebin.com/fRYDT0CF

**Comment by bogdan-iancu** [comment]
@bcnewlin , just to confirm, this crash happened while running with QM_MALLOC + DBG_MALLOC, right ? I'm asking as this coredump was not triggered by the memory debugger.

Anyhow, in frame 0, could you do `p ((struct tcp_data*)c->proto_data)` ?

**Comment by bcnewlin** [comment]
@bogdan-iancu Yes, you can see the flags set in the version info that I got from the same node.

```
(gdb) p ((struct tcp_data*)c->proto_data)
$1 = (struct tcp_data *) 0x7f2491a4cef8
```

**Comment by bcnewlin** [comment]
That was the core from the process that OpenSIPS logs said had crashed. But there actually were a couple of other cores dumped just before that.

https://pastebin.com/xx7H9Rnd
https://pastebin.com/jvwn4xMf

**Comment by bogdan-iancu** [comment]
@bcnewlin, in regards to the latest comment : the first paste link says "404 Not Found". The second one shows a totally different bracktrace for the crash, somewhere in libssl, when trying to start a curl query.

Now, going back to the gdb exploring, could you do `p *((struct tcp_data*)c->proto_data)` ?

**Comment by bcnewlin** [comment]
I don't know what is going on with pastebin, I definitely posted that first crash but now it is gone.

"totally different bracktrace for the crash, somewhere in libssl, when trying to start a curl query."
Yes, isn't that the signature of these crashes? The name of this issue was even updated to reflect that. The 2 simultaneous crashes in libcrypto is what we've been seeing a lot.

**Comment by bogdan-iancu** [comment]
@bcnewlin , reviewing all the backtraces posted here and in #2258 , I see two actual problems/crashes:
1) one in the TCP layer, when trying to flush the pending async data on a SIP TCP conn - https://pastebin.com/fRYDT0CF
2) one in the libssl, when doing an async rest post - https://pastebin.com/jvwn4xMf

These 2 crashes are repeating and not sure if related or not. So I suggest to split them over 2 different tickets - let's keep this for (2), as originally reported and move (1) into #2258

**Comment by bcnewlin** [comment]
Here is a repost of the missing backtrace link above: https://pastebin.com/3e8T1aZb

**Comment by bcnewlin** [comment]
@bogdan-iancu The backtraces may have different issues but it seems to me they must be related. Both of the backtraces you mention (one for this issue and one for #2258 happened on the same instance, only milliseconds apart from each other. The libssl/curl crashes were first.

**Comment by bogdan-iancu** [comment]
OK, so they always come in pairs with async rest post being all the time first. Maybe the #2258 is a side effect after the crash reported here. So, we should focus more on this ticket, right ?
In relation to the crash here, do you see any errors in the opensips logs (regarding maybe the SSL part)?

**Comment by bcnewlin** [comment]
Here are the logs leading up to the crash: https://pastebin.com/7G4nuw34

There are some SSL cert failures, but those certs are expected to be invalid; it's a test. Also, those are for SIP TLS connections, not HTTPS.

**Comment by bogdan-iancu** [comment]
@bcnewlin , could you run your tests, but **without** coming across the TLS verification error ? I'm trying to see if there is any relation between the 
```
 ERROR:proto_tls:tls_print_errstack: TLS errstack: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
```
error and the crash ?

**Comment by bcnewlin** [comment]
We can try to stop that specific test and see if it reproduces, but I'm not sure if we can stop only that one and not the rest of the legit TLS tests, which may also be required.

**Comment by bogdan-iancu** [comment]
@bcnewlin, trying to narrow down the culprit scenario (yes, to be honest, we do not have much of a clue on what is triggering the crash), I have two suggestions:
1. as already said, try to remove from your tests the one with the certificate validation - just to see if the handling of that error impacts at all
2. could you use the curl library (used by rest_client) compiled with GNUTLS and not OpenSSL ? The idea is to eliminate any cross influence between  the SIP stack and rest_client when comes to using/re-using the openssl layer. For example, on debian, you already have libcurl with gnutls, so it may be an easy one (remove libcurl with openssl, install the one with gnutls, rebuild the rest_client module)
```
libcurl3-gnutls - easy-to-use client-side URL transfer library (GnuTLS flavour)
libcurl3-nss - easy-to-use client-side URL transfer library (NSS flavour)
libcurl4 - easy-to-use client-side URL transfer library (OpenSSL flavour)
```

**Comment by volga629-1** [comment]
This seems happened in 3.1 too I reported yesterday. Opensips can't start  when rest client is loaded in async mode with httpS.
Our API link is contain valid certificates and no warning.
If I comment out rest client async code, opensips starts no problem 
#2292 
 
```
[root@voice-proxy ~]# rpm -qa | grep curl
libcurl-7.69.1-6.fc32.x86_64
curl-7.69.1-6.fc32.x86_64
```
Module params

```
#### Rest client
#loadmodule "rest_client.so"
#modparam("rest_client", "curl_http_version", 2)
#modparam("rest_client", "ssl_verifyhost", 0)
```

**Comment by bogdan-iancu** [comment]
@volga629-1 , I think it's a different issue.

**Comment by bcnewlin** [comment]
@bogdan-iancu 
1. We have disabled all of our "negative" TLS testing and have not had any crashes in a week. So it does seem clear that the error is related to handling of "bad" TLS scenarios like invalid certs. We are going to start re-enabling the failure tests one at a time to see when the crashes start.
2. This is difficult as on amazon linux 2 there are not extensive different packages available. We would need to compile our own version of libcurl using gnu-tls. If this is purely a troubleshooting effort, it is a big time investment for that. If the suggestion was to use that going forward, that also gives me pause as Amazon curates their packages carefully for amazon linux 2 and replacing them with our own versions could have consequences that may not be immediately obvious. Also, we would lose amazon support for that package.

**Comment by bcnewlin** [comment]
@bogdan-iancu We had another crash today, but had not re-enabled any of the "negative" TLS tests just yet. So it seems those are not completely exonerated, but may only exacerbate the issue. Unfortunately, we also did not have the extra memory debugging enabled in this build either. I'm providing the backtraces in case they are useful, and I will get the debugging turned back on.

https://pastebin.com/sqDXXhXJ
https://pastebin.com/28FNSTNf

Logs:
```
Oct 23 19:54:25 ip-10-32-30-80 kernel: opensips[7457]: segfault at 11 ip 00007f29709945c3 sp 00007ffda9bd5b60 error 4 in libcrypto.so.1.0.2k[7f2970873000+229000]
Oct 23 19:54:25 ip-10-32-30-80 kernel: opensips[7473]: segfault at 11 ip 00007f29709945c3 sp 00007ffda9bd5aa0 error 4 in libcrypto.so.1.0.2k[7f2970873000+229000]
```
version: opensips 2.4.8 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: a88a57479
main.c compiled on 17:58:28 Oct 23 2020 with gcc 7
```

**Comment by bogdan-iancu** [comment]
@bcnewlin , so right now you use SIP with TLS (with no invalid certificate) and REST with SSL, still generating crashes. Could you move the REST queries from HTTPS to HTTP (no SSL) ? Again, I running some steps in order to try to find the trigger of the crash. As it is right now, there is no much info, especially that the BT indicates somewhere deep into the libcrypto and difficult to investigate.
SO, could you test SIP+TLS with REST+HTTP (no SSL) ?

**Comment by bcnewlin** [comment]
@bogdan-iancu Unfortunately not. The calls are to an external service that does not support HTTP, only HTTPS.

**Comment by bogdan-iancu** [comment]
OK, can you somehow separate the SSL used in SIP from the one from REST ? if REST + HTTPS is a must, what about eliminating the TLS from SIP ?

BTW, for the REST w/o SSL testing, maybe a small HTTP to HTTPS proxy :D ?

**Comment by bcnewlin** [comment]
We will investigate using an HTTP proxy, but it will take some time. Is there any other way to get more information in the meantime?

**Comment by bogdan-iancu** [comment]
unfortunately no :( , as we have no clear idea what is causing this crash - more or less we are groping here.

**Comment by bcnewlin** [comment]
@bogdan-iancu we have implemented the HTTPS proxy, but are still seeing crashes. An update was provided on the wrong ticket I think on #2321 . We are going to try to enable memory debugging and reproduce.

**Comment by bogdan-iancu** [comment]
@bcnewlin , in the latest git for 2.4 (upcoming 2.4.10), there are some essential fixes related to TLS. I would strongly advice to first upgrade to latest 2.4, before doing any other troubleshooting - there are high chances to have the issue already fixed.

**Comment by bcnewlin** [comment]
@bogdan-iancu Unfortunately we have already updated and the crashes are still occurring, though I am not sure whether they are these crashes or the ones in #2321 . We are still investigating.
